src/inmanta/data/schema.py

```diff
@@ -88,159 +88,178 @@ def __init__(self, name: str, package: ModuleType, connection: Connection) -> No
         self.logger.info("Migrating from old schema management to new schema management")
+        if all_versions is None:
+            all_versions = [version.version for version in self._get_update_functions()]
         # tx begin
         async with self.connection.transaction():
-            # lock legacy table => if gone -> continue
+            # lock table
+            await self.connection.execute(f"LOCK TABLE {SCHEMA_VERSION_TABLE} IN ACCESS EXCLUSIVE MODE")
+            # get_legacy_version, under lock => if column gone -> continue
             try:
-                await self.connection.execute(f"LOCK TABLE {LEGACY_SCHEMA_VERSION_TABLE} IN ACCESS EXCLUSIVE MODE")
-            except UndefinedTableError:
-                self.logger.info("Second process is preforming a database update as well.")
+                await self.get_legacy_version()
+            except ColumnNotFound:
+                self.logger.info("Second process is performing a database update as well.")
                 return
-            # get_legacy_version, under lock
-            legacy_version_db_schema = await self.get_legacy_version()
-
-            if legacy_version_db_schema > 0:
-                self.logger.info("Creating schema version table and setting core version to %d", legacy_version_db_schema)
-                # create table
-                await self.connection.execute(create_schemamanager)
-                await self.connection.execute(
-                    f"INSERT INTO {SCHEMA_VERSION_TABLE}(name, current_version) VALUES ($1, $2)",
-                    CORE_SCHEMA_NAME,
-                    legacy_version_db_schema,
-                )
-            else:
-                self.logger.info("Creating schema version table")
-                # create table
-                await self.connection.execute(create_schemamanager)
 
-            await self.connection.execute(f"DROP TABLE {LEGACY_SCHEMA_VERSION_TABLE}")
+            self.logger.info(
+                "Creating installed_versions column in schema version table and setting installed versions based on current"
+                " legacy version."
+            )
+            await self.connection.execute(
+                f"""
+                ALTER TABLE {SCHEMA_VERSION_TABLE}
+                ADD COLUMN installed_versions integer[]
+                """
+            )
+            records: List[Record] = await self.connection.fetch(
+                f"""
+                SELECT name, current_version
+                FROM {SCHEMA_VERSION_TABLE}
+                """
+            )
+            await self.connection.executemany(
+                f"""
+                UPDATE {SCHEMA_VERSION_TABLE}
+                SET installed_versions=$1
+                WHERE name=$2
+                """,
+                (
+                    (list(takewhile(lambda x: x <= record["current_version"], all_versions)), record["name"])
+                    for record in records
+                ),
+            )
```

Comment:
This is indeed a bug, thanks. I wanted to make this a `classmethod` but then it couldn't access the connection object.

Reply:
Do you have a suggestion on how to approach this? The best approach I see would be to only update the schema for `self` after all, but then we would need to keep the old column around and we couldn't add the `NOT NULL` constraint to the new one. That wouldn't be very clean.

Reply:
Indeed that's not clean. Is there any reason why we want to store all previously applied updates. Isn't it sufficient to just keep track of the latest version at the time of the migration. All new version numbers will be a lot larger than the ones we currently have. No?

@wouterdb Any ideas/opinions?

Reply:
But then what would be the logic to select which versions need to be applied when doing an update?

Reply:
Only apply the logic on anything `>=20210101` and assume everthing lower than that has been dealt with according to the old scheme?

Reply:
That wouldn't work if there are currently versions on iso4 that aren't on iso3 (versions that are still following the old schema).

Reply:
I think that's a bit dangerous. I was thinking to apply everything `>=<latest_version_at_time_of_migration>`. Basically we would keep the old column in the database, but we would need to rename it as such that it better explains its purpose.

Reply:
I see, I still think it's a bit weird but it is infinitely better than what I proposed above. I have no objections.

Reply:
On second thought, I think there are some issues with it: suppose we do a patch `20210401` on `iso3`. Then, when migrating to `iso4`, the legacy update will be triggered and nothing below `20210311` will be applied. This could be partially mitigated by keeping to the old version numbering on `iso3`, but we'd still need to be careful not to exceed the lowest consecutive version on `iso4` (`v17` iirc). Or we could just rename `v17 -> v<today>`.

Reply:
No. The latest consecutive version on iso4 is `v7`

Reply:
"consecutive" wasn't the appropriate word. I meant `v17` is the first version on iso4 that is not on iso3. If we make sure to keep following the old version numbering on iso3 that gives us the ability to apply 9 patches, which should probably suffice. If we rename `v17` to a datestamp-based version we create more room still. In either case, we need to make sure not to add any datestamped versions on iso3 if we follow this approach.

Reply:
No. The first version in iso4 that is not in iso3 is `v7`.

Reply:
Oh, because `v7` is disabled on iso3 (I had only listed the files before, not looked at their contents)? I just realized we can't rename `v7` and `v17` either, because they'd be reapplied when updating. So I'm not sure we can easily constrain update application to `>=<latest_version_at_time_of_migration>`.

Reply:
I would keep the old collumn around and use that to determine on a per-extension basis if a migration is required.

Reply:
After thinking it through some more I think I made some minor reasoning errors when comparing both approaches. As I understand it now, @wouterdb's approach would be more or less the same as @arnaudsjs', with the addition that the legacy versions still get backfilled in the new column. I think this allows for a simpler migration away from the two columns when the time comes because only the legacy update will need to be touched, not the actual version update logic. I'm going to go with that if there are no objections.

Reply:
No objections.